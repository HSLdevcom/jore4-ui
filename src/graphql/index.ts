export * from './auth';
export * from './client';
export * from './infrastructureNetwork';
// export all values from journey patterns when it has named exports
// eslint-disable-next-line import/export
export * from './journeyPattern';
export * from './route';
export * from './servicePattern';
export * from './types';
