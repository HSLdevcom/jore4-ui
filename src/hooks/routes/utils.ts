export const extractFirstAndLastStopFromStops = (
  stopIdsWithinRoute: UUID[],
) => {
  // TODO: These will be removed from schema, remove from here as well
  const startingStopId = stopIdsWithinRoute[0];
  const finalStopId = stopIdsWithinRoute[stopIdsWithinRoute.length - 1];

  return { startingStopId, finalStopId };
};
