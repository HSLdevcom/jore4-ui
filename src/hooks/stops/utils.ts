import uniqBy from 'lodash/uniqBy';
import {
  JourneyPatternJourneyPattern,
  ServicePatternScheduledStopPoint,
} from '../../generated/graphql';

// checking whether this stop is the start or end stop of an existing route
export const isStartingOrEndingStopOfAnyRoute = (
  stopWithRouteGraphData: ServicePatternScheduledStopPoint,
) => {
  const stopId = stopWithRouteGraphData.scheduled_stop_point_id;
  return stopWithRouteGraphData.scheduled_stop_point_in_journey_patterns.some(
    (item) => {
      // journey patterns/routes that this stop is part of
      const route = item.journey_pattern.journey_pattern_route;
      // is the stop the start or end stop of this given route
      return (
        stopId === route?.starts_from_scheduled_stop_point_id ||
        stopId === route?.ends_at_scheduled_stop_point_id
      );
    },
  );
};

// gets the unique list of parent routes for the input journey patterns
export const getRoutesOfJourneyPatterns = (
  journeyPatterns: JourneyPatternJourneyPattern[],
) => {
  const allRoutes = journeyPatterns
    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
    .map((item) => item.journey_pattern_route!);

  // in the future, multiple journey patterns may have the same route,
  // so let's make sure we only return unique results
  return uniqBy(allRoutes, (route) => route.route_id);
};
