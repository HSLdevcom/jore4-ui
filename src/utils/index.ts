export * from './forms';
export * from './url';
