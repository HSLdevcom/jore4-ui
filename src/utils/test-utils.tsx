// Based on https://testing-library.com/docs/react-testing-library/setup#custom-render

import { render, RenderOptions } from '@testing-library/react'; // eslint-disable-line import/no-extraneous-dependencies
import React, { FunctionComponent } from 'react';

const AllTheProviders: FunctionComponent = ({ children }) => {
  // Add "providers" or "wrappers" that are needed in all DOM render tests here
  return <>{children}</>;
};

const customRender = (
  ui: React.ReactElement,
  options?: Omit<RenderOptions, 'queries'>,
) => render(ui, { wrapper: AllTheProviders, ...options });

// re-export everything
// eslint-disable-next-line import/no-extraneous-dependencies
export * from '@testing-library/react';
// override render method
export { customRender as render };
