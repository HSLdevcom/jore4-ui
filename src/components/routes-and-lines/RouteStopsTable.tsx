import React from 'react';
import { RouteRoute } from '../../generated/graphql';
import { RoutesTableRow } from '../RoutesTableRow'; // eslint-disable-line import/no-cycle

interface Props {
  className?: string;
  route: RouteRoute;
  testId?: string;
}

export const RouteStopsTable = ({
  className,
  route,
  testId,
}: Props): JSX.Element => {
  // TODO reusing another component to display route info; should be replaced with custom component
  return (
    <table className={`w-full ${className}`} data-testid={testId}>
      <tbody>
        <RoutesTableRow key={route.route_id} route={route} />
      </tbody>
    </table>
  );
};
