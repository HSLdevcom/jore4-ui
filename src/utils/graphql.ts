import { ApolloCache, Reference, StoreObject } from '@apollo/client';
import { Scalars } from '../generated/graphql';
import { Point } from '../types';

export const mapToObject = (object: ExplicitAny) => {
  return { object };
};

export const mapToVariables = (variables: ExplicitAny) => {
  return { variables };
};

export const mapPointToPointGeography = ({
  latitude,
  longitude,
}: Point): Scalars['geography'] => {
  // TODO: where should we get z-coordinate? Api schema requires it.
  // Use 0 as z-coordinate for now.
  return { type: 'Point', coordinates: [longitude, latitude, 0] };
};

// Removes item from apollo's cache.
// TODO: do we really have to do this manually?
export const removeFromApolloCache = (
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  cache: ApolloCache<any>,
  identity: StoreObject | Reference,
) => {
  // Based on https://stackoverflow.com/a/66713628
  const cached = cache.identify(identity);
  // @ts-expect-error something seems to be wrong here. The solution mentioned
  // above stackoverflow response won't give ts errors, but it doesn't work either...
  cache.evict(cached);
  cache.gc();
};
