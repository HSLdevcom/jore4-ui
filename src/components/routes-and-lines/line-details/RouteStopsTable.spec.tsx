import { fireEvent, screen } from '@testing-library/react';
import { DateTime } from 'luxon';
import React from 'react';
import { GetLineDetailsWithRoutesByIdQuery } from '../../../generated/graphql';
import {
  GqlQueryResult,
  mapLineDetailsWithRoutesResult,
} from '../../../graphql';
import { buildLocalizedString, render } from '../../../utils/test-utils';
import { RouteStopsTable } from './RouteStopsTable';

describe(`<${RouteStopsTable.name} />`, () => {
  const testId = 'routeStopsTable1';

  // this response is copy-pasted from the actual graphql response
  const mockResponse = {
    data: {
      route_line_by_pk: {
        line_id: '101f800c-39ed-4d85-8ece-187cd9fe1c5e',
        name_i18n: buildLocalizedString('Rautatientori - Veräjälaakso'),
        short_name_i18n: buildLocalizedString('Rautatientori - Veräjälaakso'),
        primary_vehicle_mode: 'bus',
        validity_start: DateTime.fromISO('2021-01-01T00:00:00+00:00'),
        validity_end: DateTime.fromISO('2023-12-13T00:00:00+00:00'),
        priority: 10,
        label: '65',
        __typename: 'route_line',
        line_routes: [
          {
            route_id: '03d55414-e5cf-4cce-9faf-d86ccb7e5f98',
            description_i18n: null,
            starts_from_scheduled_stop_point_id:
              'e3528755-711f-4e4f-9461-7931a2c4bc6d',
            ends_at_scheduled_stop_point_id:
              'f8eace87-7901-4438-bfee-bb6f24f1c4c4',
            route_shape: {
              type: 'LineString',
              coordinates: [
                [24.927431159, 60.163633535, 8.809],
                [24.929195906, 60.164250755, 8.621],
                [24.932010801, 60.165237492, 10.901],
                [24.934523102, 60.166118268, 15.564],
              ],
            },
            on_line_id: '101f800c-39ed-4d85-8ece-187cd9fe1c5e',
            validity_start: DateTime.fromISO('2021-01-01T00:00:00+00:00'),
            validity_end: DateTime.fromISO('2023-12-13T00:00:00+00:00'),
            priority: 10,
            label: '65 itään',
            direction: 'outbound',
            __typename: 'route_route',
            starts_from_scheduled_stop_point: {
              scheduled_stop_point_id: 'e3528755-711f-4e4f-9461-7931a2c4bc6d',
              label: 'pysäkki A',
              validity_start: DateTime.fromISO('2021-01-01T00:00:00+00:00'),
              validity_end: DateTime.fromISO('2023-12-13T00:00:00+00:00'),
              __typename: 'service_pattern_scheduled_stop_point',
            },
            ends_at_scheduled_stop_point: {
              scheduled_stop_point_id: 'f8eace87-7901-4438-bfee-bb6f24f1c4c4',
              label: 'pysäkki C',
              validity_start: DateTime.fromISO('2021-01-01T00:00:00+00:00'),
              validity_end: DateTime.fromISO('2023-12-13T00:00:00+00:00'),
              __typename: 'service_pattern_scheduled_stop_point',
            },
            infrastructure_links_along_route: [
              {
                infrastructure_link: {
                  scheduled_stop_point_located_on_infrastructure_link: [
                    {
                      scheduled_stop_point_id:
                        'e3528755-711f-4e4f-9461-7931a2c4bc6d',
                      label: 'pysäkki A',
                      validity_start: DateTime.fromISO(
                        '2021-01-01T00:00:00+00:00',
                      ),
                      validity_end: DateTime.fromISO(
                        '2023-12-13T00:00:00+00:00',
                      ),
                      __typename: 'service_pattern_scheduled_stop_point',
                      scheduled_stop_point_in_journey_patterns: [
                        {
                          journey_pattern_id:
                            '43e1985d-4643-4415-8367-c4a37fbc0a87',
                          scheduled_stop_point_id:
                            'e3528755-711f-4e4f-9461-7931a2c4bc6d',
                          scheduled_stop_point_sequence: 0,
                          is_timing_point: false,
                          is_via_point: false,
                          __typename:
                            'journey_pattern_scheduled_stop_point_in_journey_pattern',
                        },
                      ],
                    },
                  ],
                  __typename: 'infrastructure_network_infrastructure_link',
                },
                __typename: 'route_infrastructure_link_along_route',
              },
              {
                infrastructure_link: {
                  scheduled_stop_point_located_on_infrastructure_link: [
                    {
                      scheduled_stop_point_id:
                        '4d294d62-df17-46ff-9248-23f66f17fa87',
                      label: 'pysäkki B',
                      validity_start: DateTime.fromISO(
                        '2021-01-01T00:00:00+00:00',
                      ),
                      validity_end: DateTime.fromISO(
                        '2023-12-13T00:00:00+00:00',
                      ),
                      __typename: 'service_pattern_scheduled_stop_point',
                      scheduled_stop_point_in_journey_patterns: [
                        {
                          journey_pattern_id:
                            '43e1985d-4643-4415-8367-c4a37fbc0a87',
                          scheduled_stop_point_id:
                            '4d294d62-df17-46ff-9248-23f66f17fa87',
                          scheduled_stop_point_sequence: 1,
                          is_timing_point: true,
                          is_via_point: true,
                          __typename:
                            'journey_pattern_scheduled_stop_point_in_journey_pattern',
                        },
                      ],
                    },
                  ],
                  __typename: 'infrastructure_network_infrastructure_link',
                },
                __typename: 'route_infrastructure_link_along_route',
              },
              {
                infrastructure_link: {
                  scheduled_stop_point_located_on_infrastructure_link: [
                    {
                      scheduled_stop_point_id:
                        'f8eace87-7901-4438-bfee-bb6f24f1c4c4',
                      label: 'pysäkki C',
                      validity_start: DateTime.fromISO(
                        '2021-01-01T00:00:00+00:00',
                      ),
                      validity_end: DateTime.fromISO(
                        '2023-12-13T00:00:00+00:00',
                      ),
                      __typename: 'service_pattern_scheduled_stop_point',
                      scheduled_stop_point_in_journey_patterns: [
                        {
                          journey_pattern_id:
                            '43e1985d-4643-4415-8367-c4a37fbc0a87',
                          scheduled_stop_point_id:
                            'f8eace87-7901-4438-bfee-bb6f24f1c4c4',
                          scheduled_stop_point_sequence: 2,
                          is_timing_point: false,
                          is_via_point: false,
                          __typename:
                            'journey_pattern_scheduled_stop_point_in_journey_pattern',
                        },
                      ],
                    },
                  ],
                  __typename: 'infrastructure_network_infrastructure_link',
                },
                __typename: 'route_infrastructure_link_along_route',
              },
            ],
          },
        ],
      },
    },
  } as GqlQueryResult<GetLineDetailsWithRoutesByIdQuery>;

  test('Renders the route with stops along its geometry', async () => {
    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
    const line = mapLineDetailsWithRoutesResult(mockResponse)!;
    const { container, asFragment } = render(
      <RouteStopsTable testId={testId} routes={line.line_routes} />,
    );

    // the stops don't show as the accordion is not open
    expect(container.querySelectorAll('tr')).toHaveLength(1);
    expect(asFragment()).toMatchSnapshot();

    const accordionButton = screen.getByTestId(
      'RouteStopsHeaderRow::toggleAccordion',
    );
    fireEvent.click(accordionButton);

    // the stops should show when the accordion opens
    expect(container.querySelectorAll('tr')).toHaveLength(4);
    expect(asFragment()).toMatchSnapshot();
  });
});
