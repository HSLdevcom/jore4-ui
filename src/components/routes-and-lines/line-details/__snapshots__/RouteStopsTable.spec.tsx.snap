// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RouteStopsTable /> Renders the route with stops along its geometry 1`] = `
<DocumentFragment>
  <div>
    <div
      class="flex items-center"
    >
      <label
        class="my-8 mr-8 font-normal"
        id="headlessui-label-1"
      >
        Näytä reittiin kuulumattomat pysäkit
      </label>
      <button
        aria-checked="true"
        aria-labelledby="headlessui-label-1"
        class="undefined border-brand bg-brand relative inline-flex h-6 w-11 items-center rounded-full border transition-colors"
        id="headlessui-switch-2"
        role="switch"
        tabindex="0"
        type="button"
      >
        <span
          class="translate-x-5 border-brand inline-block h-6 w-6 transform rounded-full border bg-white transition-transform"
        />
      </button>
    </div>
    <table
      class="h-1 w-full undefined"
      data-testid="routeStopsTable1"
    >
      <tbody>
        <tr
          class="border border-white bg-background undefined"
        >
          <td
            class="border-l-8 py-4 pl-16 pr-4 text-3xl font-bold"
          >
            <div
              class="flex flex-row items-center"
            >
              <span
                class="relative mr-4 flex h-12 w-12 items-center justify-center bg-brand text-white"
              >
                <span
                  data-testid="direction-value"
                >
                  1
                </span>
                <i
                  class="icon-opposite border-brank absolute -bottom-2 -right-2 rounded-full border bg-white text-sm text-brand"
                />
              </span>
              <span
                data-testid="route-row-label"
              >
                65 itään
              </span>
            </div>
          </td>
          <td
            class="flex items-center py-4 text-3xl"
          >
            <span
              data-testid="route-row-name"
            />
            <a
              data-testid="edit-route-button"
              href="/routes/03d55414-e5cf-4cce-9faf-d86ccb7e5f98/edit"
            >
              <div
                class="ml-5 flex h-10 w-10 items-center justify-center rounded-full border border-grey bg-white hover:border-2 hover:border-tweaked-brand"
              >
                <svg
                  class="text-2xl text-tweaked-brand"
                  fill="currentColor"
                  height="1em"
                  stroke="currentColor"
                  stroke-width="0"
                  viewBox="0 0 24 24"
                  width="1em"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M0 0h24v24H0z"
                    fill="none"
                  />
                  <path
                    d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75L3 17.25zM20.71 7.04a.996.996 0 000-1.41l-2.34-2.34a.996.996 0 00-1.41 0l-1.83 1.83 3.75 3.75 1.83-1.83z"
                  />
                </svg>
              </div>
            </a>
          </td>
          <td
            class="pr-16 text-right"
            data-testid="route-row-validity-period"
          >
            Voimassa 1.1.2021 - 13.12.2023
          </td>
          <td>
            <div
              class="flex flex-row items-center"
            >
              <span
                data-testid="route-row-last-edited"
              >
                !14.2.2017 klo 14.51
              </span>
              <svg
                class="ml-2 inline text-xl text-tweaked-brand"
                fill="currentColor"
                height="1em"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 24 24"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M0 0h24v24H0V0z"
                  fill="none"
                />
                <path
                  d="M13 3a9 9 0 00-9 9H1l3.89 3.89.07.14L9 12H6c0-3.87 3.13-7 7-7s7 3.13 7 7-3.13 7-7 7c-1.93 0-3.68-.79-4.94-2.06l-1.42 1.42A8.954 8.954 0 0013 21a9 9 0 000-18zm-1 5v5l4.25 2.52.77-1.28-3.52-2.09V8z"
                />
              </svg>
            </div>
          </td>
          <td
            class="w-20 border-l-4 border-r-4 border-white text-center"
          >
            <button
              class="text-center h-10 w-10 rounded-full border border-grey bg-white text-tweaked-brand hover:border-2 hover:border-tweaked-brand "
              data-testid="RouteStopsHeaderRow::showRoute"
              type="button"
            >
              <svg
                class="text-2xl inline text-center"
                fill="currentColor"
                height="1em"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 24 24"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M0 0h24v24H0z"
                  fill="none"
                />
                <path
                  d="M18 8c0-3.31-2.69-6-6-6S6 4.69 6 8c0 4.5 6 11 6 11s6-6.5 6-11zm-8 0c0-1.1.9-2 2-2s2 .9 2 2a2 2 0 11-4 0zM5 20v2h14v-2H5z"
                />
              </svg>
            </button>
          </td>
          <td
            class="w-20"
          >
            <button
              class="h-full w-full text-center"
              data-testid="RouteStopsHeaderRow::toggleAccordion"
              type="button"
            >
              <svg
                class="inline text-center text-3xl text-tweaked-brand"
                fill="currentColor"
                height="1em"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 448 512"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M207.029 381.476L12.686 187.132c-9.373-9.373-9.373-24.569 0-33.941l22.667-22.667c9.357-9.357 24.522-9.375 33.901-.04L224 284.505l154.745-154.021c9.379-9.335 24.544-9.317 33.901.04l22.667 22.667c9.373 9.373 9.373 24.569 0 33.941L240.971 381.476c-9.373 9.372-24.569 9.372-33.942 0z"
                />
              </svg>
            </button>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</DocumentFragment>
`;

exports[`<RouteStopsTable /> Renders the route with stops along its geometry 2`] = `
<DocumentFragment>
  <div>
    <div
      class="flex items-center"
    >
      <label
        class="my-8 mr-8 font-normal"
        id="headlessui-label-1"
      >
        Näytä reittiin kuulumattomat pysäkit
      </label>
      <button
        aria-checked="true"
        aria-labelledby="headlessui-label-1"
        class="undefined border-brand bg-brand relative inline-flex h-6 w-11 items-center rounded-full border transition-colors"
        id="headlessui-switch-2"
        role="switch"
        tabindex="0"
        type="button"
      >
        <span
          class="translate-x-5 border-brand inline-block h-6 w-6 transform rounded-full border bg-white transition-transform"
        />
      </button>
    </div>
    <table
      class="h-1 w-full undefined"
      data-testid="routeStopsTable1"
    >
      <tbody>
        <tr
          class="border border-white bg-background undefined"
        >
          <td
            class="border-l-8 py-4 pl-16 pr-4 text-3xl font-bold"
          >
            <div
              class="flex flex-row items-center"
            >
              <span
                class="relative mr-4 flex h-12 w-12 items-center justify-center bg-brand text-white"
              >
                <span
                  data-testid="direction-value"
                >
                  1
                </span>
                <i
                  class="icon-opposite border-brank absolute -bottom-2 -right-2 rounded-full border bg-white text-sm text-brand"
                />
              </span>
              <span
                data-testid="route-row-label"
              >
                65 itään
              </span>
            </div>
          </td>
          <td
            class="flex items-center py-4 text-3xl"
          >
            <span
              data-testid="route-row-name"
            />
            <a
              data-testid="edit-route-button"
              href="/routes/03d55414-e5cf-4cce-9faf-d86ccb7e5f98/edit"
            >
              <div
                class="ml-5 flex h-10 w-10 items-center justify-center rounded-full border border-grey bg-white hover:border-2 hover:border-tweaked-brand"
              >
                <svg
                  class="text-2xl text-tweaked-brand"
                  fill="currentColor"
                  height="1em"
                  stroke="currentColor"
                  stroke-width="0"
                  viewBox="0 0 24 24"
                  width="1em"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M0 0h24v24H0z"
                    fill="none"
                  />
                  <path
                    d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75L3 17.25zM20.71 7.04a.996.996 0 000-1.41l-2.34-2.34a.996.996 0 00-1.41 0l-1.83 1.83 3.75 3.75 1.83-1.83z"
                  />
                </svg>
              </div>
            </a>
          </td>
          <td
            class="pr-16 text-right"
            data-testid="route-row-validity-period"
          >
            Voimassa 1.1.2021 - 13.12.2023
          </td>
          <td>
            <div
              class="flex flex-row items-center"
            >
              <span
                data-testid="route-row-last-edited"
              >
                !14.2.2017 klo 14.51
              </span>
              <svg
                class="ml-2 inline text-xl text-tweaked-brand"
                fill="currentColor"
                height="1em"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 24 24"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M0 0h24v24H0V0z"
                  fill="none"
                />
                <path
                  d="M13 3a9 9 0 00-9 9H1l3.89 3.89.07.14L9 12H6c0-3.87 3.13-7 7-7s7 3.13 7 7-3.13 7-7 7c-1.93 0-3.68-.79-4.94-2.06l-1.42 1.42A8.954 8.954 0 0013 21a9 9 0 000-18zm-1 5v5l4.25 2.52.77-1.28-3.52-2.09V8z"
                />
              </svg>
            </div>
          </td>
          <td
            class="w-20 border-l-4 border-r-4 border-white text-center"
          >
            <button
              class="text-center h-10 w-10 rounded-full border border-grey bg-white text-tweaked-brand hover:border-2 hover:border-tweaked-brand "
              data-testid="RouteStopsHeaderRow::showRoute"
              type="button"
            >
              <svg
                class="text-2xl inline text-center"
                fill="currentColor"
                height="1em"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 24 24"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M0 0h24v24H0z"
                  fill="none"
                />
                <path
                  d="M18 8c0-3.31-2.69-6-6-6S6 4.69 6 8c0 4.5 6 11 6 11s6-6.5 6-11zm-8 0c0-1.1.9-2 2-2s2 .9 2 2a2 2 0 11-4 0zM5 20v2h14v-2H5z"
                />
              </svg>
            </button>
          </td>
          <td
            class="w-20"
          >
            <button
              class="h-full w-full text-center"
              data-testid="RouteStopsHeaderRow::toggleAccordion"
              type="button"
            >
              <svg
                class="inline text-center text-3xl text-tweaked-brand"
                fill="currentColor"
                height="1em"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 448 512"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M240.971 130.524l194.343 194.343c9.373 9.373 9.373 24.569 0 33.941l-22.667 22.667c-9.357 9.357-24.522 9.375-33.901.04L224 227.495 69.255 381.516c-9.379 9.335-24.544 9.317-33.901-.04l-22.667-22.667c-9.373-9.373-9.373-24.569 0-33.941L207.03 130.525c9.372-9.373 24.568-9.373 33.941-.001z"
                />
              </svg>
            </button>
          </td>
        </tr>
        <tr
          class="border border-l-8 bg-background text-dark-grey undefined"
        >
          <td
            class="py-4 pl-16 pr-4"
            data-testid="stop-row-label"
          >
            pysäkki A
          </td>
          <td
            data-testid="stop-row-name"
          >
            !Pysäkki X
          </td>
          <td
            class="pr-16 text-right"
            data-testid="stop-row-validity-period"
          >
            Ei reitin käytössä
          </td>
          <td>
            <div
              class="flex flex-row items-center"
            >
              <span
                data-testid="stop-row-last-edited"
              >
                !14.2.2017 klo 14.51
              </span>
              <svg
                class="ml-2 inline text-xl text-tweaked-brand"
                fill="currentColor"
                height="1em"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 24 24"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M0 0h24v24H0V0z"
                  fill="none"
                />
                <path
                  d="M13 3a9 9 0 00-9 9H1l3.89 3.89.07.14L9 12H6c0-3.87 3.13-7 7-7s7 3.13 7 7-3.13 7-7 7c-1.93 0-3.68-.79-4.94-2.06l-1.42 1.42A8.954 8.954 0 0013 21a9 9 0 000-18zm-1 5v5l4.25 2.52.77-1.28-3.52-2.09V8z"
                />
              </svg>
            </div>
          </td>
          <td>
             
          </td>
          <td>
            <div
              class="relative"
            >
              <button
                aria-expanded="false"
                aria-haspopup="true"
                class="mx-auto flex items-center px-3 focus:outline-none"
                data-testid="stop-row-action-menu"
                id="headlessui-menu-button-3"
                type="button"
              >
                <svg
                  class="text-3xl"
                  fill="currentColor"
                  height="1em"
                  stroke="currentColor"
                  stroke-width="0"
                  viewBox="0 0 24 24"
                  width="1em"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M0 0h24v24H0z"
                    fill="none"
                  />
                  <path
                    d="M12 8c1.1 0 2-.9 2-2s-.9-2-2-2-2 .9-2 2 .9 2 2 2zm0 2c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2zm0 6c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2z"
                  />
                </svg>
              </button>
            </div>
          </td>
        </tr>
        <tr
          class="border border-l-8 bg-background text-dark-grey undefined"
        >
          <td
            class="py-4 pl-16 pr-4"
            data-testid="stop-row-label"
          >
            pysäkki B
          </td>
          <td
            data-testid="stop-row-name"
          >
            !Pysäkki X
          </td>
          <td
            class="pr-16 text-right"
            data-testid="stop-row-validity-period"
          >
            Ei reitin käytössä
          </td>
          <td>
            <div
              class="flex flex-row items-center"
            >
              <span
                data-testid="stop-row-last-edited"
              >
                !14.2.2017 klo 14.51
              </span>
              <svg
                class="ml-2 inline text-xl text-tweaked-brand"
                fill="currentColor"
                height="1em"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 24 24"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M0 0h24v24H0V0z"
                  fill="none"
                />
                <path
                  d="M13 3a9 9 0 00-9 9H1l3.89 3.89.07.14L9 12H6c0-3.87 3.13-7 7-7s7 3.13 7 7-3.13 7-7 7c-1.93 0-3.68-.79-4.94-2.06l-1.42 1.42A8.954 8.954 0 0013 21a9 9 0 000-18zm-1 5v5l4.25 2.52.77-1.28-3.52-2.09V8z"
                />
              </svg>
            </div>
          </td>
          <td>
             
          </td>
          <td>
            <div
              class="relative"
            >
              <button
                aria-expanded="false"
                aria-haspopup="true"
                class="mx-auto flex items-center px-3 focus:outline-none"
                data-testid="stop-row-action-menu"
                id="headlessui-menu-button-4"
                type="button"
              >
                <svg
                  class="text-3xl"
                  fill="currentColor"
                  height="1em"
                  stroke="currentColor"
                  stroke-width="0"
                  viewBox="0 0 24 24"
                  width="1em"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M0 0h24v24H0z"
                    fill="none"
                  />
                  <path
                    d="M12 8c1.1 0 2-.9 2-2s-.9-2-2-2-2 .9-2 2 .9 2 2 2zm0 2c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2zm0 6c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2z"
                  />
                </svg>
              </button>
            </div>
          </td>
        </tr>
        <tr
          class="border border-l-8 bg-background text-dark-grey undefined"
        >
          <td
            class="py-4 pl-16 pr-4"
            data-testid="stop-row-label"
          >
            pysäkki C
          </td>
          <td
            data-testid="stop-row-name"
          >
            !Pysäkki X
          </td>
          <td
            class="pr-16 text-right"
            data-testid="stop-row-validity-period"
          >
            Ei reitin käytössä
          </td>
          <td>
            <div
              class="flex flex-row items-center"
            >
              <span
                data-testid="stop-row-last-edited"
              >
                !14.2.2017 klo 14.51
              </span>
              <svg
                class="ml-2 inline text-xl text-tweaked-brand"
                fill="currentColor"
                height="1em"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 24 24"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M0 0h24v24H0V0z"
                  fill="none"
                />
                <path
                  d="M13 3a9 9 0 00-9 9H1l3.89 3.89.07.14L9 12H6c0-3.87 3.13-7 7-7s7 3.13 7 7-3.13 7-7 7c-1.93 0-3.68-.79-4.94-2.06l-1.42 1.42A8.954 8.954 0 0013 21a9 9 0 000-18zm-1 5v5l4.25 2.52.77-1.28-3.52-2.09V8z"
                />
              </svg>
            </div>
          </td>
          <td>
             
          </td>
          <td>
            <div
              class="relative"
            >
              <button
                aria-expanded="false"
                aria-haspopup="true"
                class="mx-auto flex items-center px-3 focus:outline-none"
                data-testid="stop-row-action-menu"
                id="headlessui-menu-button-5"
                type="button"
              >
                <svg
                  class="text-3xl"
                  fill="currentColor"
                  height="1em"
                  stroke="currentColor"
                  stroke-width="0"
                  viewBox="0 0 24 24"
                  width="1em"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M0 0h24v24H0z"
                    fill="none"
                  />
                  <path
                    d="M12 8c1.1 0 2-.9 2-2s-.9-2-2-2-2 .9-2 2 .9 2 2 2zm0 2c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2zm0 6c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2z"
                  />
                </svg>
              </button>
            </div>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</DocumentFragment>
`;
